clr/commands.py

```diff
@@ -421,28 +421,55 @@ class NamespaceCache:
 
     This allows subsequent calls to `clr help` or `clr completion_*` to be fast.
     Necessary to work around the fact that many clr command namespace modules
-    import the world and initialize state on import.
+    import the world and initialize state on import. The cache is used only for
+    help and completion calls, when actually calling a clr command its spec is
+    always loaded from the module.
     """
 
     def __init__(self):
+        # Place the cache file in the temp dir with a static file name. Data stored in
+        # this file is easily regeneratable in the event of a system restart, but should
+        # be left behind after the process is complete for subsequent clr calls to find.
         tmpdir = os.environ.get("TMPDIR", "/tmp")
         self.cache_fn = os.path.join(tmpdir, "clr_command_cache")
-        # Clr processes are short lived. We don't close the shelve, but are
-        # careful to sync it after writes.
-        self.cache = shelve.open(self.cache_fn)
+
+        try:
+            # Cache is stored on disk as a shelve, but loaded into memory and then
+            # closed right away so that multiple processes won't conflict trying to
+            # access it.
+            with shelve.open(self.cache_fn) as cache_shelve:
+                self.cache = dict(cache_shelve)
+        except Exception as e:
+            # Caching is considered best effort and fails silently. Can always load the
+            # module, this is just slower.
+            self.cache = {}
 
     def get(self, namespace_key):
         # Don't cache the system namespace. It is already loaded.
         if namespace_key == "system":
             return get_namespace("system")
-        # Load namespace and save spec to the shelve.
-        if namespace_key not in self.cache:
-            namespace = get_namespace(namespace_key)
-            if isinstance(namespace, ErrorLoadingNamespace):
-                return namespace
-            self.cache[namespace_key] = NamespaceCacheEntry.create(namespace)
-            self.cache.sync()
-        return self.cache[namespace_key]
+        # Return from cache if present.
+        if namespace_key in self.cache:
+            return self.cache[namespace_key]
+        return self.load_and_sync_entry(namespace_key)
+
+    def load_and_sync_entry(self, namespace_key):
+        namespace = get_namespace(namespace_key)
+        if isinstance(namespace, ErrorLoadingNamespace):
+            # Don't cache errors.
+            return namespace
+
+        entry = NamespaceCacheEntry.create(namespace)
+        self.cache[namespace_key] = entry
+
+        # Try to save the entry to disk. Fail silently.
+        try:
+            with shelve.open(self.cache_fn) as cache_shelve:
+                cache_shelve[namespace_key] = entry
+        except Exception as e:
```

Comment:
Same as above, can we:
- not assign exception instance to `e` if it will be unused
- catch specific classes of error, not base `Exception`
- log the failure (if necessary)

Reply:
My thinking for this was that the cache is purely an optimization detail of the clr help and completion subcommands, which still work completely on failure. While it would be a shame if they somehow were always failing and no one noticed, I'm more concerned about them complaining and being a distraction. If the cache was 100% broken we will notice because tab completions will have a ~1second delay.